File: cpp/src/ilqrsolver.cpp
#include "ilqrsolver.h"

/* Debug */
#include <iostream>
using namespace std;
/* */

using namespace Eigen;

ILQRSolver::ILQRSolver(DynamicModel& myDynamicModel, CostFunction& myCostFunction,bool fullDDP,bool QPBox)
{
    dynamicModel = &myDynamicModel;
    costFunction = &myCostFunction;
    stateNb = myDynamicModel.getStateNb();
    commandNb = myDynamicModel.getCommandNb();
    enableQPBox = QPBox;
    enableFullDDP = fullDDP;
    if(QPBox)
    {
        qp = new QProblemB(commandNb);
        Options myOptions;
        myOptions.printLevel = PL_LOW;
        myOptions.enableRegularisation = BT_TRUE;
        myOptions.initialStatusBounds = ST_INACTIVE;
        myOptions.numRefinementSteps = 1;
        myOptions.enableCholeskyRefactorisation = 1;
        qp->setOptions(myOptions);

        xOpt = new real_t[commandNb];
        lowerCommandBounds = myDynamicModel.getLowerCommandBounds();
        upperCommandBounds = myDynamicModel.getUpperCommandBounds();
    }
}

void ILQRSolver::FirstInitSolver(stateVec_t& myxInit, stateVec_t& myxDes, unsigned int& myT,
                       double& mydt, unsigned int& myiterMax,double& mystopCrit)
{
    xInit = myxInit-myxDes;
    xDes = myxDes;
    T = myT;
    dt = mydt;
    iterMax = myiterMax;
    stopCrit = mystopCrit;

    xList.resize(myT+1);
    uList.resize(myT);
    updatedxList.resize(myT+1);
    updateduList.resize(myT);
    tmpxPtr.resize(myT+1);
    tmpuPtr.resize(myT);
    k.setZero();
    K.setZero();
    kList.resize(myT);
    KList.resize(myT);

    /*xList = new stateVec_t[myT+1];
    uList = new commandVec_t[myT];
    updatedxList = new stateVec_t[myT+1];
    updateduList = new commandVec_t[myT];
    k.setZero();
    K.setZero();
    kList = new commandVec_t[myT];
    KList = new commandR_stateC_t[myT];*/

    alphaList[0] = 1.0;
    alphaList[1] = 0.8;
    alphaList[2] = 0.6;
    alphaList[3] = 0.4;
    alphaList[4] = 0.2;
    alpha = 1.0;
}

void ILQRSolver::initSolver(stateVec_t& myxInit, stateVec_t& myxDes)
{
    xInit = myxInit - myxDes;
    xDes = myxDes;
}

void ILQRSolver::solveTrajectory()
{
    initTrajectory();
    for(iter=0;iter<iterMax;iter++)
    {
        backwardLoop();
        forwardLoop();
        if(changeAmount<stopCrit)
        {
          break;
        }
        tmpxPtr = xList;
        tmpuPtr = uList;
        xList = updatedxList;
        updatedxList = tmpxPtr;
        uList = updateduList;
        updateduList = tmpuPtr;
    }
}

void ILQRSolver::initTrajectory()
{
    xList[0] = xInit;
    commandVec_t zeroCommand;
    stateVec_t Xe;
    Xe = xInit;
    zeroCommand.setZero();
    for(unsigned int i=0;i<T;i++)
    {
        uList[i] = zeroCommand;
        xList[i+1] = dynamicModel->computeNextState(dt,xList[i] ,xDes,zeroCommand) - xDes;
        //Xe = xList[i] - xDes;
    }
}

void ILQRSolver::backwardLoop()
{
    costFunction->computeFinalCostDeriv(xList[T]);
    nextVx = costFunction->getlx();
    nextVxx = costFunction->getlxx();

    mu = 0.0;
    completeBackwardFlag = 0;

    while(!completeBackwardFlag)
    {
        completeBackwardFlag = 1;
        muEye = mu*stateMat_t::Zero();
        for(int i=T-1;i>=0;i--)
        {
            x = xList[i];
            u = uList[i];

            dynamicModel->computeAllModelDeriv(dt,x,xDes,u);
            costFunction->computeAllCostDeriv(x,u);

            Qx = costFunction->getlx() + dynamicModel->getfx().transpose() * nextVx;
            Qu = costFunction->getlu() + dynamicModel->getfu().transpose() * nextVx;
            Qxx = costFunction->getlxx() + dynamicModel->getfx().transpose() * (nextVxx+muEye) * dynamicModel->getfx();
            Quu = costFunction->getluu() + dynamicModel->getfu().transpose() * (nextVxx+muEye) * dynamicModel->getfu();
            Qux = costFunction->getlux() + dynamicModel->getfu().transpose() * (nextVxx+muEye) * dynamicModel->getfx();

            if(enableFullDDP)
            {
                Qxx += dynamicModel->computeTensorContxx(nextVx);
                Qux += dynamicModel->computeTensorContux(nextVx);
                Quu += dynamicModel->computeTensorContuu(nextVx);
            }

            QuuInv = Quu.inverse();

            if(!isQuudefinitePositive(Quu))
            {
                /*
                  To be Implemented : Regularization (is Quu definite positive ?)
                */
                if(mu==0.0) mu += 1e-4;
                else mu *= 10;
                completeBackwardFlag = 0;
                break;
            }

            if(enableQPBox)
            {
                nWSR = 10;
                H = Quu;
                g = Qu;
                lb = lowerCommandBounds - u;
                ub = upperCommandBounds - u;
                qp->init(H.data(),g.data(),lb.data(),ub.data(),nWSR);
                qp->getPrimalSolution(xOpt);
                k = Map<commandVec_t>(xOpt);
                K = -QuuInv*Qux;
                for(unsigned int i_cmd=0;i_cmd<commandNb;i_cmd++)
                {
                    if((k[i_cmd] == lowerCommandBounds[i_cmd]) | (k[i_cmd] == upperCommandBounds[i_cmd]))
                    {
                        K.row(i_cmd).setZero();
                    }
                }
            }
            else
            {
                k = -QuuInv*Qu;
                K = -QuuInv*Qux;
            }

            /*nextVx = Qx - K.transpose()*Quu*k;
            nextVxx = Qxx - K.transpose()*Quu*K;*/
            nextVx = Qx + K.transpose()*Quu*k + K.transpose()*Qu + Qux.transpose()*k;
            nextVxx = Qxx + K.transpose()*Quu*K+ K.transpose()*Qux + Qux.transpose()*K;
            nextVxx = 0.5*(nextVxx + nextVxx.transpose());

            kList[i] = k;
            KList[i] = K;
        }
    }
}

void ILQRSolver::forwardLoop()
{
    changeAmount = 0.0;
    updatedxList[0] = xInit;
    // Line search to be implemented
    alpha = 1.0;
    for(unsigned int i=0;i<T;i++)
    {
        updateduList[i] = uList[i] + alpha*kList[i] + KList[i]*(updatedxList[i] - xList[i]);
        updatedxList[i+1] = dynamicModel->computeNextState(dt,updatedxList[i],xDes,updateduList[i]) - xDes;
        for(unsigned int j=0;j<commandNb;j++)
        {
            changeAmount += abs(uList[i](j,0) - updateduList[i](j,0));
        }
    }
}

ILQRSolver::traj ILQRSolver::getLastSolvedTrajectory()
{
    lastTraj.xList = updatedxList;
    for(int i=0;i<T+1;i++)lastTraj.xList[i] += xDes;
    lastTraj.uList = updateduList;
    for(int i=0;i<T;i++) lastTraj.uList[i] += dynamicModel->commandOffset;
    lastTraj.iter = iter;
    return lastTraj;
}

bool ILQRSolver::isQuudefinitePositive(const commandMat_t & Quu)
{
    /*
      Todo : check if Quu is definite positive
    */
    //Eigen::JacobiSVD<commandMat_t> svd_Quu (Quu, ComputeThinU | ComputeThinV);
    Eigen::VectorXcd singular_values = Quu.eigenvalues();

    for(long i = 0; i < Quu.cols(); ++i)
    {
        if (singular_values[i].real() < 0.)
        {
            std::cout << "not sdp" << std::endl;
            return false;
        }
    }
    return true;
}

File: python/iLQRSolver.py
import numpy as np
import numpy.linalg
import matplotlib.pyplot as pl
from costFunction import *
from dynamicModel import *
import time

class ILQRSolver:
    def __init__(self,model,costFunction):
        self.model = model
        self.costfunction = costFunction
        ''' init values changed when solveTrajectory is called '''
        self.xinit = np.zeros((model.stateNumber,1))
        self.Xdes = np.zeros((model.stateNumber,1))
        self.T = 10
        self.dt = 1e-4
        self.iterMax = 20
        self.stopCrit = 1e-3
        ''' '''
        self.changeAmount = 0.0


    def solveTrajectory(self,Xinit,Xdes,T,dt,iterMax=20,stopCrit=1e-3):
        """
        _ initialisation
        _ backward Pass
        _ optimization of regularization coef
        _ forward Pass with optimization of line-search coef
        _ stop when convergence or iterMax reached
        """
        self.Xinit = Xinit
        self.Xdes = Xdes
        self.T = T
        self.dt = dt
        self.iterMax = iterMax
        self.stopCrit = stopCrit

        XList,UList = self.initTrajectory()
        for iter in range(self.iterMax):
            kList,KList = self.backwardLoop(XList,UList)
            newXList,newUList = self.forwardLoop(kList,KList,XList,UList)
            XList = newXList
            UList = newUList

            if(self.changeAmount < self.stopCrit):
                break

        return XList,UList,iter


    def initTrajectory(self):
        XList = []
        append = XList.append
        computeNextState = self.model.computeNextState
        zerosCommand = np.zeros((self.model.commandNumber,1))
        dt = self.dt
        UList = [zerosCommand for i in range(self.T)]
        append(self.Xinit)
        for i in range(self.T):
            append(computeNextState(dt,XList[i],UList[i]))
        return XList,UList


    def computeAllDeriv(self,X,Xdes,U):
        ''' useless '''
        fx,fxx,fu,fuu,fxu,fux = self.model.computeAllModelDeriv(self.dt,X,U)
        lx,lxx,lu,luu,lxu,lux = self.costfunction.computeAllCostDeriv(X,Xdes,U)
        return fx,fxx,fu,fuu,fxu,fux,lx,lxx,lu,luu,lxu,lux


    def backwardLoop(self,XList,UList):
        computeAllModelDeriv = self.model.computeAllModelDeriv
        computeAllCostDeriv = self.costfunction.computeAllCostDeriv
        kList = []
        KList = []
        kappend = kList.append
        Kappend = KList.append
        nextVx,nextVxx = \
            self.costfunction.computeFinalCostDeriv(XList[self.T],self.Xdes)

        Xdes = self.Xdes
        stateNb = self.model.stateNumber
        dt = self.dt

        mu = 0.0
        completeBackwardFlag = 0
        while(completeBackwardFlag==0):
            completeBackwardFlag = 1
            muEye = mu*np.eye( nextVxx.shape[0] , dtype=float)
            for i in range(self.T-1,-1,-1):  # backward pass
                X = XList[i]
                U = UList[i]

                fx,fxx,fu,fuu,fxu,fux = computeAllModelDeriv(dt,X,U)
                lx,lxx,lu,luu,lxu,lux = computeAllCostDeriv(X,Xdes,U)

                Qx = lx + fx.T*nextVx
                Qu = lu + fu.T*nextVx
                Qxx = lxx + fx.T*nextVxx*fx
                # Quu = luu + fu.T*nextVxx*fu
                # Qux = lux + fu.T*nextVxx*fx
                # Qxu = lxu + fx.T*nextVxx*fu

                Quut = luu + fu.T*(nextVxx+muEye)*fu
                Quxt = lux + fu.T*(nextVxx+muEye)*fx
                for j in range(stateNb): # second order derivative of dynamic
                    Qxx += nextVx[j].item()*fxx[j]
                    # Qux += nextVx[j].item()*fux[j]
                    Quxt += nextVx[j].item()*fux[j]
                    # Quu += nextVx[j].item()*fuu[j]
                    Quut += nextVx[j].item()*fuu[j]


                if(np.all(np.linalg.eigvals(Quut) <= 0)): # check out if Quut is definite positive
                    if(mu==0):
                        mu += 1e-4
                    else:
                        mu = mu*10
                    completeBackwardFlag = 0
                    break

                QuuInv = np.linalg.inv(Quut)

                k = - QuuInv*Qu
                K = - QuuInv*Quxt

                # regularization (Y. Tassa thesis)
                # nextVx      = Qx + K.T*Quut*k + K.T*Qu + Quxt.T*k
                # nextVxx     = Qxx + K.T*Quut*K + K.T*Quxt + Quxt.T*K
                nextVx = Qx - K.T*Quut*k
                nextVxx = Qxx - K.T*Quut*K

                kappend(k)
                Kappend(K)
                # kList[i] = k
                # KList[i] = K
        kList.reverse()
        KList.reverse()
        return kList,KList

    def forwardLoop(self,kList,KList,XList,UList):
        newXList = []
        newUList = []
        xappend = newXList.append
        uappend = newUList.append
        computeNextState = self.model.computeNextState
        changeAmount = 0.0
        dt = self.dt
        commandNb = self.model.commandNumber

        xappend(self.Xinit)
        # Line search to be implemented
        alphaList = [1.0,0.8,0.6,0.4,0.2]
        alpha = alphaList[0]
        for i in range(self.T):
            uappend( UList[i] + alpha*kList[i] \
                        + KList[i]*(newXList[i] - XList[i]))
            xappend(computeNextState(dt,newXList[i],newUList[i]))
            for j in range(commandNb):
                U = UList[i]
                newU = newUList[i]
                changeAmount += np.abs(U[j] - newU[j])
        self.changeAmount = changeAmount
        return newXList,newUList